### PR TITLE
EZP-31745: Introduce PagerSearchDataMapper interface

### PR DESCRIPTION
**Pull request by ITernovtsii**
| Question      | Answer
| ------------- | ---
| Tickets       |  [EZP-31745](https://jira.ez.no/browse/EZP-31745)
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Tests pass?   | no
| Doc needed?   | no
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-search/blob/master/LICENSE)

Currently, if I want to add some extra information in search result template, I can decorate
"Ibexa\Platform\Search\Mapper\PagerSearchContentToDataMapper" service.
But, decorator service is not accepted by SearchViewBuilder.

#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)

### CODE REVIEW

File: src/bundle/Resources/config/services.yaml [hunk 1]
```diff
@@ -1,25 +1,28 @@
...
-          $searchService: '@ezpublish.api.service.search'
+    Ibexa\Platform\Search\Mapper\PagerSearchContentToDataMapper:
+        arguments:
+            $contentTypeService: '@ezpublish.api.service.content_type'
+            $userService: '@ezpublish.api.service.user'
+            $userLanguagePreferenceProvider: '@eZ\Publish\Core\MVC\Symfony\Locale\UserLanguagePreferenceProvider'
+            $translationHelper: '@ezpublish.translation_helper'
+            $languageService: '@ezpublish.api.service.language'
+
+    Ibexa\Platform\Search\Mapper\PagerDataMapperInterface: '@Ibexa\Platform\Search\Mapper\PagerSearchContentToDataMapper'
```
Comment (ViniTou) [inline src/bundle/Resources/config/services.yaml:24]: Could we not tie those two together? As I understand those are for autowiring but it should work without it as concrete implementation is specified in service definition.

File: src/lib/Mapper/PagerDataMapperInterface.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace Ibexa\Platform\Search\Mapper;
+
+use Pagerfanta\Pagerfanta;
+
+interface PagerDataMapperInterface
+{
+    public function map(Pagerfanta $pagerfanta);
```
Comment (ViniTou) [inline src/lib/Mapper/PagerDataMapperInterface.php:13]: could we add return typehint (`array`) as well?

**Review by ViniTou — CHANGES_REQUESTED** [review]

**Comment by ViniTou** [comment]
Also please rebase it and target it against 3.1 branch.

File: src/bundle/Resources/config/services.yaml (hunk 1 above)
Comment (ITernovtsiy) [reply src/bundle/Resources/config/services.yaml:24]: No, it will lead to this error:
```
Cannot autowire service "App\Controller\AlertController": argument "$pagerDataMapper" of method "__construct()" references interface "Ibexa\Platform\Search\Mapper\PagerDataMapperInterface" but no such service exists. You should maybe alias this interface to one of these existing services: "EzSystems\EzPlatformAdminUi\Search\PagerSearchContentToDataMapper", "Ibexa\Platform\Search\Mapper\PagerSearchContentToDataMapper".
```
I will open one more PR to admin-ui bundle to include:
`Ibexa\Platform\Search\Mapper\PagerDataMapperInterface: '@EzSystems\EzPlatformAdminUi\Search\PagerSearchContentToDataMapper'`
Ok?

Comment (ViniTou) [reply src/bundle/Resources/config/services.yaml:24]: Then that code
`    Ibexa\Platform\Search\Mapper\PagerDataMapperInterface: '@Ibexa\Platform\Search\Mapper\PagerSearchContentToDataMapper'
`

should either go to your App (if you use autowire), or you should inject specifying instance and not relay on it.

File: src/lib/Mapper/PagerDataMapperInterface.php (hunk 2 above)
Comment (ITernovtsiy) [reply src/lib/Mapper/PagerDataMapperInterface.php]: sure, added

**Comment by ITernovtsiy** [comment]
@ViniTou 
Sorry, I don't see 3.1 branch here https://github.com/ezsystems/ezplatform-search/branches 
I rebased it to 1.0.

**Comment by ViniTou** [comment]
@ezsystems/php-dev-team 
As I totally agree with changing typehint for mapper to some abstraction, I now have mixed feelings about placing this interface here, or maybe rather with naming. If you look now, we already have `\EzSystems\EzPlatformAdminUi\Tab\Dashboard\PagerContentToDataMapper` which on first glance should also implement this interface but on other it is completely unrelated to Search.

**Review by kaff — APPROVED** [review]

**Review by alongosz — CHANGES_REQUESTED** [review]
### TL;DR;:
In this case I would suggest an event which will be dispatched inside the mapper to collect required extra parameters and attach them in each iteration or after the mapping, inside Search View Builder, to process an entire `$view`.

### Review:
First of all, there are some out of scope whitespace changes, which makes it difficult to review. Not sure why you need them, but if you do, place them in a separate commit, please.

> As I totally agree with changing typehint for mapper to some abstraction, I now have mixed feelings about placing this interface here, or maybe rather with naming. If you look now, we already have `\EzSystems\EzPlatformAdminUi\Tab\Dashboard\PagerContentToDataMapper` which on first glance should also implement this interface but on other it is completely unrelated to Search.

Since it uses separate namespaces, no strong feelings here. Could be `PagerSearchDataMapper` or language-wise more correct - but inconsistent - `SearchPagerDataMapper`. _Notice that I didn't use Interface suffix here intentionally, because I consider it a bad extension of Hungarian notation and one of the worst practices advocated both by PSR and Symfony, for reasons passing my understanding._

Naming aside, using decoration here is very error prone and breaks a bit responsibility, because decorator would be required to provide all the necessary data items, including the core ones. This will lead to possible BC breaks, when our mapper gets updated with some new required parameters. We cannot give a BC promise on that, otherwise updating wouldn't be possible.

**Comment by ITernovtsiy** [comment]
> First of all, there are some out of scope whitespace changes, which makes it difficult to review. Not sure why you need them, but if you do, place them in a separate commit, please.

Ok, fix for yaml code style removed.

> Since it uses separate namespaces, no strong feelings here. Could be PagerSearchDataMapper or language-wise more correct - but inconsistent - SearchPagerDataMapper. Notice that I didn't use Interface suffix here intentionally, because I consider it a bad extension of Hungarian notation and one of the worst practices advocated both by PSR and Symfony, for reasons passing my understanding.

Agree about naming) updated to "PagerSearchDataMapper".

> using decoration here is very error prone and breaks a bit responsibility, because decorator would be required to provide all the necessary data items, including the core ones. This will lead to possible BC breaks, when our mapper gets updated with some new required parameters. We cannot give a BC promise on that, otherwise updating wouldn't be possible.

Sorry, but I didn't find any better way of extending this.
I used this in `services.yaml`
```
ContextualCode\EzPlatformAdminSearchBundle\Mapper\PagerSearchContentToDataMapper:
decorates: Ibexa\Platform\Search\Mapper\PagerSearchContentToDataMapper
```
and define class as:
```
use Ibexa\Platform\Search\Mapper\PagerSearchContentToDataMapper as BaseMapper;

class PagerSearchContentToDataMapper implements PagerSearchDataMapper
{
    /** @var BaseMapper */
    private $baseMapper;
    /** @var LocationService */
    private $locationService;
    /** @var BookmarkService */
    private $bookmarkService;

    public function __construct(
        BaseMapper $baseMapper,
        BookmarkService $bookmarkService,
        LocationService $locationService
    ) {
        $this->baseMapper = $baseMapper;
        $this->locationService = $locationService;
        $this->bookmarkService = $bookmarkService;
    }

    public function map(Pagerfanta $pager): array
    {
        $data = $this->baseMapper->map($pager);

        foreach ($data as $i => $result) {
            $data[$i]['location_is_bookmarked'] = $this->bookmarkService->isBookmarked(
                $this->locationService->loadLocation($result['mainLocationId'])
            );
        }

        return $data;
    }
}
```
This will work if the core mapper updated with some new parameters. _It will break if `mainLocationId` removed from the core mapper, but I can live with this..._
If you have any suggestions about how to better extend the mapper, I would appreciate it.

**Review by ViniTou — APPROVED** [review]

File: src/lib/Mapper/PagerSearchDataMapper.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace Ibexa\Platform\Search\Mapper;
```
Comment (alongosz) [inline src/lib/Mapper/PagerSearchDataMapper.php:7]:
```php
declare(strict_types=1);
```

File: src/lib/Mapper/PagerSearchContentToSearchDataMapper.php [hunk 4]
```diff
@@ -20,7 +20,7 @@
 use eZ\Publish\Core\MVC\Symfony\Locale\UserLanguagePreferenceProviderInterface;
 use Pagerfanta\Pagerfanta;
 
-class PagerSearchContentToDataMapper
+class PagerSearchContentToSearchDataMapper implements PagerSearchDataMapper
```
Comment (alongosz) [inline src/lib/Mapper/PagerSearchContentToSearchDataMapper.php:23]: This is a BC break for this package. Any code which relied on `PagerSearchContentToDataMapper` is going to stop working. What you could do is to re-add this class in the form of `PagerSearchContentToDataMapper extends PagerSearchContentToSearchDataMapper`  and annotate it as `@deprecated since Ibexa v3.2, use PagerSearchContentToSearchDataMapper instead`

**Review by alongosz — CHANGES_REQUESTED** [review]
In general 1.0 is no longer supported, rebase to 1.1 and fix other issues please.
